stdlib/src/builtin/string.mojo

```diff
@@ -1408,6 +1408,35 @@ struct String(
             substr._strref_dangerous(), start=start
         )
 
+    fn isspace(self) -> Bool:
+        """Determines whether the given String is a python
+        whitespace String. This corresponds to Python's
+        [universal separators](
+            https://docs.python.org/3/library/stdtypes.html#str.splitlines)
+        `" \\t\\n\\r\\f\\v\\x1c\\x1e\\x85\\u2028\\u2029"`.
+
+        Returns:
+            True if the String is one of the whitespace characters
+                listed above, otherwise False."""
+        # TODO add line and paragraph separator as stringliteral
+        # once unicode escape secuences are accepted
+        # Unicode Line Separator: \ u2028
+        # Unicode Paragraph Separator: \ u2029
+        # 0 is to build a String with null terminator
+        return self in List[String](
+            String(" "),
+            String("\t"),
+            String("\n"),
+            String("\r"),
+            String("\v"),
+            String("\f"),
+            String("\x1c"),
+            String("\x1e"),
+            String("\x85"),
+            String(List[UInt8](0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x38, 0)),
+            String(List[UInt8](0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x39, 0)),
+        )
```

Comment:
Could this work? It would guarantee that the list is built at compile time. If it doesn't work, you could also try some collection of `StringLiteral`.
```suggestion
        alias spaces = List[String](
            String(" "),
            String("\t"),
            String("\n"),
            String("\r"),
            String("\v"),
            String("\f"),
            String("\x1c"),
            String("\x1e"),
            String("\x85"),
            String(List[UInt8](0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x38, 0)),
            String(List[UInt8](0x20, 0x5C, 0x75, 0x32, 0x30, 0x32, 0x39, 0)),
        )
        return self in spaces
```
Or better yet, if the alias could be pushed into a private alias on the struct, you could ensure in the tests that its contents don't accidentally change.

Reply:
> compile time initalization of the list of space characters

What I fear might happen is that making it an alias concatenates all the chars in the list (that's why I didn't do it at first)

![image](https://github.com/modularml/mojo/assets/110240700/b5c7eaeb-02c7-4d6e-a1a5-e5d5a9008584)

I will try it and make a test for it to see if some of the bytes that compose a unicode separator get counted as being in the alias. I think we need to find of a better way to deal with different UTF standards...

> a private alias on the struct

That's also another topic that I wanted to talk about. Currently many other methods use String.WHITESPACE, in my `String.split()` PR I'm adding utf8 chars that are whitespace into that variable, but many others should also take unicode separators into account and they don't. If we change it to a `List[String]` and refactor those functions to iterate 4 bytes at a time, we will pay a performance penalty but we will be more conformant with Python. WDYT?

> Could you do this for StringRef, and then just delegate to that? 

I thought by looking at the code that we wanted to have most complex string operations to be on string.mojo (ord, atol, char, etc. are all there)

> the top-level isspace function could be removed

That function behaves like C's `isspace()` so maybe renaming it? It will be much faster than `String.isspace()`since it will be checking for less chars and only in a StringRef directly with `memcmp`

Reply:
> > a private alias on the struct
> 
> That's also another topic that I wanted to talk about. Currently many other methods use String.WHITESPACE, in my `String.split()` PR I'm adding utf8 chars that are whitespace into that variable, but many others should also take unicode separators into account and they don't. If we change it to a `List[String]` and refactor those functions to iterate 4 bytes at a time, we will pay a performance penalty but we will be more conformant with Python. WDYT?

We can extend Python's API with either parameter or argument flags to control this behavior. For now, I wouldn't worry too much about performance and try to get to the functionality first.

> 
> > Could you do this for StringRef, and then just delegate to that?
> 
> I thought by looking at the code that we wanted to have most complex string operations to be on string.mojo (ord, atol, char, etc. are all there)

Fair point, and `StringRef` should also mostly go away in favor of `StringSlice. We would probably need a similar method on that, though.

> 
> > the top-level isspace function could be removed
> 
> That function behaves like C's `isspace()` so maybe renaming it? It will be much faster than `String.isspace()`since it will be checking for less chars and only in a StringRef directly with `memcmp`

I checked, `String.isspace` is actually just an implementation detail of `String` and `StringRef`, so we should just make it private.